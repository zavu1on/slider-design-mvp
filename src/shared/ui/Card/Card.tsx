import type { FC, ReactNode } from 'react';
import {
  CardContent,
  CardDescription,
  CardFooter,
  CardHeader,
  Card as CardSkeleton,
  CardTitle,
} from '@/generated/shadcn/card';

type CardProps = {
  children: ReactNode;
  title: string;
  description?: string;
  footer?: ReactNode;
  className?: string;
  onClick?: () => void;

  style?: Partial<{
    header: string;
    title: string;
    description: string;
    content: string;
    footer: string;
  }>;
};

export const Card: FC<CardProps> = ({
  children,
  title,
  description,
  footer,
  className,
  onClick,
  style,
}) => {
  return (
    <CardSkeleton className={className} onClick={onClick}>
      <CardHeader className={style?.header}>
        <CardTitle className={style?.title}>{title}</CardTitle>
        <CardDescription className={style?.description}>
          {description}
        </CardDescription>
      </CardHeader>
      <CardContent className={style?.content}>{children}</CardContent>
      <CardFooter className={style?.footer}>{footer}</CardFooter>
    </CardSkeleton>
  );
};
