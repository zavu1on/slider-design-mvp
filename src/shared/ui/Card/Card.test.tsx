import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import { afterEach, describe, expect, it, vi } from 'vitest';
import { Card } from './Card';

describe('Card component', () => {
  afterEach(() => {
    cleanup();
  });

  it('renders correctly with title and description (snapshot)', () => {
    const { container } = render(
      <Card title="Test title" description="Test description">
        <p>Card content</p>
      </Card>
    );

    expect(container).toMatchSnapshot();
  });

  it('renders children inside CardContent', () => {
    render(
      <Card title="Card with children">
        <span>Body text</span>
      </Card>
    );

    const content = screen.getByText('Body text');
    expect(content.closest('[data-slot="card-content"]')).toBeInTheDocument();
  });

  it('renders footer when provided', () => {
    render(
      <Card title="Card with footer" footer={<button>Click me</button>}>
        Content
      </Card>
    );

    expect(screen.getByTestId('card-footer')).toBeInTheDocument();
    expect(screen.getByText('Click me')).toBeInTheDocument();
  });

  it('applies custom className to root CardSkeleton', () => {
    render(
      <Card title="Styled card" className="custom-class">
        Content
      </Card>
    );
    const root = screen.getByTestId('card-container');
    expect(root).toHaveClass('custom-class');
  });

  it('applies style overrides for sub-elements', () => {
    render(
      <Card
        title="Styled"
        description="Styled description"
        style={{
          header: 'header-class',
          title: 'title-class',
          description: 'desc-class',
          content: 'content-class',
          footer: 'footer-class',
        }}
        footer={<span>Footer</span>}
      >
        Body
      </Card>
    );

    expect(screen.getByText('Styled').className).toContain('title-class');
    expect(screen.getByText('Styled description').className).toContain(
      'desc-class'
    );
    expect(screen.getByText('Body').className).toContain('content-class');
    expect(screen.getByTestId('card-footer').className).toContain(
      'footer-class'
    );
  });

  it('calls onClick handler when clicked', () => {
    const handleClick = vi.fn();
    render(
      <Card title="Clickable card" onClick={handleClick}>
        Body
      </Card>
    );

    fireEvent.click(screen.getByTestId('card-container'));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('matches snapshot with all props provided', () => {
    const { container } = render(
      <Card
        title="Full card"
        description="Detailed description"
        footer={<button>Footer btn</button>}
        className="outer-class"
        onClick={() => {}}
        style={{
          header: 'header-x',
          title: 'title-x',
          description: 'desc-x',
          content: 'content-x',
          footer: 'footer-x',
        }}
      >
        <div>Full body</div>
      </Card>
    );

    expect(container).toMatchSnapshot();
  });
});
