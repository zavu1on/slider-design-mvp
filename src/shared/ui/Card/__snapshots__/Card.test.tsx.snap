// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card component > matches snapshot with all props provided 1`] = `
<div>
  <div
    class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm outer-class"
    data-slot="card"
    data-testid="card-container"
  >
    <div
      class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 header-x"
      data-slot="card-header"
    >
      <div
        class="leading-none font-semibold title-x"
        data-slot="card-title"
      >
        Full card
      </div>
      <div
        class="text-muted-foreground text-sm desc-x"
        data-slot="card-description"
      >
        Detailed description
      </div>
    </div>
    <div
      class="px-6 content-x"
      data-slot="card-content"
    >
      <div>
        Full body
      </div>
    </div>
    <div
      class="flex items-center px-6 [.border-t]:pt-6 footer-x"
      data-slot="card-footer"
      data-testid="card-footer"
    >
      <button>
        Footer btn
      </button>
    </div>
  </div>
</div>
`;

exports[`Card component > renders correctly with title and description (snapshot) 1`] = `
<div>
  <div
    class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
    data-slot="card"
    data-testid="card-container"
  >
    <div
      class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6"
      data-slot="card-header"
    >
      <div
        class="leading-none font-semibold"
        data-slot="card-title"
      >
        Test title
      </div>
      <div
        class="text-muted-foreground text-sm"
        data-slot="card-description"
      >
        Test description
      </div>
    </div>
    <div
      class="px-6"
      data-slot="card-content"
    >
      <p>
        Card content
      </p>
    </div>
    <div
      class="flex items-center px-6 [.border-t]:pt-6"
      data-slot="card-footer"
      data-testid="card-footer"
    />
  </div>
</div>
`;

exports[`Card component > renders with title and description (snapshot) 1`] = `
<div>
  <div
    class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
    data-slot="card"
  >
    <div
      class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6"
      data-slot="card-header"
    >
      <div
        class="leading-none font-semibold"
        data-slot="card-title"
      >
        Test title
      </div>
      <div
        class="text-muted-foreground text-sm"
        data-slot="card-description"
      >
        Test description
      </div>
    </div>
    <div
      class="px-6"
      data-slot="card-content"
    >
      <p>
        Card content
      </p>
    </div>
    <div
      class="flex items-center px-6 [.border-t]:pt-6"
      data-slot="card-footer"
    />
  </div>
</div>
`;
